### PR TITLE
Preserve times for fs::copy on Unix.

### PR DESCRIPTION
**Pull request by pitdicker**
This adds `set_times` to `File` so `fs::copy` can restore the access, modified and created times.
Also permissions are now set during file creation, as a partly fix for #26933.

I did not test the code for setting the creation time, as I do not have BSD.
It probably works as long as the creation time is earlier than what it is currently set to.
If we want to make a method like this public, we probably should add a way to keep some times the same, or set them to the current time.

There are still some things left for #26933, like copying attributes, extended attributes and ACLs.

### CODE REVIEW

File: src/libstd/sys/unix/fs.rs [hunk 1]
```diff
@@ -732,17 +791,126 @@ pub fn canonicalize(p: &Path) -> io::Result<PathBuf> {
 }
 
 pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
-    use fs::{File, set_permissions};
-    if !from.is_file() {
+    use fs::OpenOptions;
+    let mut from_opts = OpenOptions::new();
+    let mut reader = try!(from_opts.read(true).open(&from));
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: Shouldn't this be the same as `File::open`?

File: src/libstd/sys/unix/fs.rs [hunk 2]
```diff
@@ -732,17 +791,126 @@ pub fn canonicalize(p: &Path) -> io::Result<PathBuf> {
...
     }
 
-    let mut reader = try!(File::open(from));
-    let mut writer = try!(File::create(to));
-    let perm = try!(reader.metadata()).permissions();
+    let mut to_opts = OpenOptions::new();
+    let mut writer = try!(to_opts.write(true)
+                                 .create(true)
+                                 .truncate(true)
+                                 .mode(attr.perm().mode())
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: Can you create a writable version of a read-only file? (I'm just not sure if that works here)

File: src/libstd/sys/unix/fs.rs [hunk 3]
```diff
@@ -732,17 +791,126 @@ pub fn canonicalize(p: &Path) -> io::Result<PathBuf> {
...
+    pub fn tmpdir() -> TempDir {
+        let p = env::temp_dir();
+        let mut r = rand::thread_rng();
+        let ret = p.join(&format!("rust-{}", r.next_u32()));
+        check!(fs::create_dir(&ret));
+        TempDir(ret)
+    }
+
+    #[test]
+    fn copy_preserves_times() {
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: I wonder if this should be a cross-platform test? Windows should also copy file times, right?

File: src/libstd/sys/unix/fs.rs [hunk 4]
```diff
@@ -498,6 +498,65 @@ impl File {
...
+                     modified: SystemTime,
+                     created: Option<SystemTime>) -> io::Result<()> {
+        if let Some(birthtime) = created {
+            // On some (and maybe all) versions of BSD the birth time can be set
+            // thanks to the guarantee that the birth time is always equal to or
+            // earlier then the last modified time.
+            // So the birth time can be set by setting the times twice:
+            // once with times[1] set to the birth time, and then with times[1]
+            // set to the modified time.
+            try!(self.set_times_inner(accessed, birthtime));
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: Out of curiosity, is there precedent for setting the creation time when copying files?

File: src/libstd/sys/unix/fs.rs (hunk 2 above)
Comment (pitdicker) [inline src/libstd/sys/unix/fs.rs]: Yes, the permissions do not apply for the fd that created it, only to new opens.

**Comment by alexcrichton** [comment]
I'm curious if other versions of this function in other libraries also copy times by default? I wonder if this should perhaps be an opt-in behavior?

File: src/libstd/sys/unix/fs.rs (hunk 1 above)
Comment (pitdicker) [inline src/libstd/sys/unix/fs.rs]: Now I don't remember why I changed that...
If I don't find the reason I will change it back.

**Comment by pitdicker** [comment]
I would have to look, but I think it is nice to imitate cp as close as possible on Unix

File: src/libstd/sys/unix/fs.rs (hunk 3 above)
Comment (pitdicker) [inline src/libstd/sys/unix/fs.rs]: Yes and yes :). But we do not yet have a way to set file times cross-platform. Would it be better if added a way to do so?

Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: For now we probably shouldn't add functionality like that, but it's fine if the tests reaches into the `sys` modules to call various bits and pieces.

File: src/libstd/sys/unix/fs.rs (hunk 2 above)
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: Ah ok, cool! Could you add a comment to that effect?

**Comment by alexcrichton** [comment]
Ah ok, if `cp` does it, and Windows already does it that sounds like precedent enough to me!

**Comment by nodakai** [comment]
[`cp(1)`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/cp.html#tag_20_24_04) copies the original timestamp only if told
```
$ cp /etc/protocols .
$ ls -l protocols
-rw-r--r-- 1 nodakai nodakai 6.4K Mar 23 20:15 protocols
$ cp -a /etc/protocols .
$ ls -l protocols
-rw-r--r-- 1 nodakai nodakai 6.4K Jan 12  2010 protocols
```
This is probably for avoiding possible performance hit caused by inode access.

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase!

**Pull request closed without merging**